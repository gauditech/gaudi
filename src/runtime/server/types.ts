import { NextFunction, Request, Response } from "express";

export type ServerHttpMethods = "all" | "get" | "post" | "put" | "patch" | "delete";

export type ServerRequestHandler =
  | ((request: Request, response: Response) => void | Promise<void>)
  | ((req: Request, resp: Response, next: NextFunction) => void | Promise<void>);

export type EndpointConfig = {
  path: string;
  method: ServerHttpMethods;
  handlers: ServerRequestHandler[];
};

export type ServerMiddlewareNextFn = (err?: unknown) => void;
