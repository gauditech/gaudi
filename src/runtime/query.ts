import _ from "lodash";

import { Ref, RefKind, getModelProp, getRef, getTargetModel } from "@src/common/refs";
import { ensureEqual } from "@src/common/utils";
import {
  Definition,
  EndpointDef,
  FilterDef,
  IQueryDefPath,
  ModelDef,
  QueryDef,
  QueryDefPath,
  TargetDef,
} from "@src/types/definition";

type NamePath = string[];

// FIXME add tests
export function mergePaths(paths: NamePath[]): NamePath[] {
  return _.uniqWith(paths, _.isEqual);
}

export function mkContextQuery(def: Definition, endpoint: EndpointDef): QueryDef | null {
  // this is context query, drop final target
  const targets = _.initial(endpoint.targets);
  if (!targets.length) return null;
  const targetPath = targets.map((t) => t.name);
  const filterPaths = targets.flatMap((t) => getFilterPaths(targetToFilter(t)));
  const paths = mergePaths([targetPath, ...filterPaths]);
  const direct = getDirectChildren(paths);
  ensureEqual(direct.length, 1);
  const { value: ctx } = getRef<"model">(def, direct[0]);
  const joinPaths = processPaths(def, getRelatedPaths(paths, ctx.name), ctx, [ctx.name]);
  // navigate through targets to find cardinality
  const [isOne, nullable] = targets.reduce(
    (a, t) => {
      const ref = getRef<"reference" | "relation" | "query">(def, t.name);
      switch (ref.kind) {
        case "reference": {
          return [a[0], a[1] || ref.value.nullable];
        }
        case "relation": {
          return [ref.value.unique && a[0], a[1] || ref.value.nullable];
        }
        case "query": {
          return [a[0] && ref.value.retCardinality === "one", a[1] || ref.value.nullable];
        }
      }
    },
    [true, false]
  );
  return {
    name: direct[0],
    fromPath: targetPath,
    ctxModelRefKey: ctx.refKey,
    refKey: "N/A",
    nullable, // fixme
    retCardinality: isOne ? "one" : "many",
    retType: _.last(targets)!.retType,
    filter: undefined, // fixme
    joinPaths,
  };
}

export function getDirectChildren(paths: NamePath[]): string[] {
  return _.chain(paths)
    .map((p) => p[0])
    .compact()
    .uniq()
    .value();
}

export function getRelatedPaths(paths: NamePath[], direct: string): NamePath[] {
  return paths.filter((path) => path[0] === direct).map(_.tail);
}

function calculateCardinality(
  ref: Ref<"reference" | "relation" | "query">,
  paths: QueryDefPath[]
): "one" | "many" {
  const joinsAllOnes = paths.every((p) => p.retCardinality === "one");
  const isOne =
    ref.kind === "reference" ||
    (ref.kind === "relation" && ref.value.unique) ||
    (ref.kind === "query" && ref.value.retCardinality === "one");
  return isOne && joinsAllOnes ? "one" : "many";
}

export function processPaths(
  def: Definition,
  paths: NamePath[],
  parentCtx: ModelDef,
  prefixNames: NamePath
): QueryDefPath[] {
  const direct = getDirectChildren(paths);

  return _.compact(
    direct.flatMap((name): QueryDefPath | null => {
      const relativeChildren = getRelatedPaths(paths, name);
      const ref = getModelProp<"field" | "reference" | "relation" | "query">(parentCtx, name);
      if (ref.kind === "field") {
        return null;
      }

      const targetCtx = getTargetModel(def.models, ref.value.refKey);
      const joinPaths = processPaths(def, relativeChildren, targetCtx, [...prefixNames, name]);
      return {
        kind: ref.kind,
        joinType: "inner",
        refKey: ref.value.refKey,
        joinPaths,
        retType: getTargetModel(def.models, ref.value.refKey).name,
        retCardinality: calculateCardinality(ref, joinPaths),
      } as QueryDefPath;
      // FIXME typescript doesn't like this so we have to cast
    })
  );
}

function targetToFilter(target: TargetDef): FilterDef {
  return {
    kind: "binary",
    operator: "is",
    lhs: {
      kind: "alias",
      namePath: [...target.namePath, target.identifyWith.name],
    },
    rhs: { kind: "variable", name: target.identifyWith.paramName, type: target.identifyWith.type },
  };
}

// function mkTargetQuery(def: Definition, endpoint: EndpointDef): QueryDef {}

// function queryToString(def: Definition, q: QueryDef): string {}

export function getFilterPaths(filter: FilterDef): string[][] {
  switch (filter?.kind) {
    case undefined:
    case "literal":
    case "variable":
      return [];
    case "alias":
      return [[...filter.namePath]];
    case "binary": {
      return [...getFilterPaths(filter.lhs), ...getFilterPaths(filter.rhs)];
    }
  }
}
