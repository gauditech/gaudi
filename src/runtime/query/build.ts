import _ from "lodash";
import { match } from "ts-pattern";

import { kindFilter } from "@src/common/kindFilter";
import { getRef, getTargetModel } from "@src/common/refs";
import { assertUnreachable, ensureEqual } from "@src/common/utils";
import { HookCode } from "@src/types/common";
import {
  Definition,
  ModelDef,
  NestedSelect,
  QueryDef,
  QueryOrderByAtomDef,
  SelectDef,
  SelectHook,
  SelectItem,
  SelectableExpression,
  TypedExprDef,
} from "@src/types/definition";

// FIXME introduce Queryable with SelectableItem[]?
export type QueryTree = {
  name: string;
  alias: string;
  query: QueryDef;
  queryIdAlias: string | undefined;
  hooks: {
    name: string;
    args: { name: string; query: QueryTree }[];
    hook: HookCode;
  }[];
  related: QueryTree[];
};

export type NamePath = string[];

export const GAUDI_INTERNAL_TARGET_ID_ALIAS = "__gaudi__target_id";

/**
 * Utils
 */

/**
 * TODO turn this into Queryable
 *
 * Converts QueryDef into a QueryDef with only selectable expressions.
 * Ensures the `id` field is selected any non-selectable item is requeste.d
 */
function queryToQueriable(query: QueryDef): QueryDef {
  const selectables = kindFilter(query.select, "expression");
  const targetIdAlias = findTargetIdAlias(query);
  if (targetIdAlias === GAUDI_INTERNAL_TARGET_ID_ALIAS) {
    selectables.push({
      kind: "expression",
      alias: GAUDI_INTERNAL_TARGET_ID_ALIAS,
      expr: { kind: "alias", namePath: [...query.fromPath, "id"] },
      type: { kind: "integer", nullable: false },
    });
  }

  return { ...query, select: selectables };
}

function findTargetIdAlias(query: QueryDef): string | undefined {
  const selectables = kindFilter(query.select, "expression");
  const hasNested = kindFilter(query.select, "nested-select");
  const hasHooks = kindFilter(query.select, "model-hook");
  const id = selectables.find(
    (s) =>
      s.alias === "id" &&
      s.expr?.kind === "alias" &&
      _.isEqual(s.expr.namePath, [...query.fromPath, "id"])
  );
  if (id) {
    return "id";
  } else if (hasHooks || hasNested) {
    return GAUDI_INTERNAL_TARGET_ID_ALIAS;
  }
}

export function selectToHooks(select: SelectDef): SelectHook[] {
  return select.filter((s): s is SelectHook => s.kind === "model-hook");
}

export function applyFilterIdInContext(namePath: NamePath, filter?: TypedExprDef): TypedExprDef {
  const inFilter: TypedExprDef = {
    kind: "function",
    name: "in",
    args: [
      { kind: "alias", namePath: [...namePath, "id"] },
      {
        kind: "variable",
        type: { kind: "collection", type: { kind: "integer", nullable: false } },
        name: "@context_ids",
      },
    ],
  };
  return filter === undefined
    ? inFilter
    : { kind: "function", name: "and", args: [filter, inFilter] };
}

function getPathRetType(def: Definition, path: NamePath): ModelDef {
  // assume it starts with model
  const model = getRef.model(def, path[0]);
  const ret = _.tail(path).reduce(
    (ctx, name) => getTargetModel(def, `${ctx.refKey}.${name}`),
    model
  );
  return ret;
}

/**
 * Query builder
 */

export function queryTreeFromParts(
  def: Definition,
  name: string,
  fromPath: NamePath,
  filter: TypedExprDef,
  select: SelectDef
): QueryTree {
  const query = queryFromParts(def, name, fromPath, filter, select);
  const qTree = buildQueryTree(def, query);
  return qTree;
}

export function queryFromParts(
  def: Definition,
  name: string,
  fromPath: NamePath,
  filter: TypedExprDef,
  select: SelectDef,
  orderBy?: QueryOrderByAtomDef[],
  limit?: number,
  offset?: number
): QueryDef {
  const sourceModel = getRef.model(def, _.first(fromPath)!);

  return {
    kind: "query",
    refKey: "N/A",
    modelRefKey: sourceModel.refKey,
    filter,
    fromPath,
    name,
    // retCardinality: "many", // FIXME,
    retType: getPathRetType(def, fromPath).refKey,
    select,
    orderBy,
    limit,
    offset,
  };
}

export function getDirectChildren(paths: NamePath[]): string[] {
  return _.chain(paths)
    .map((p) => p[0])
    .compact()
    .uniq()
    .value();
}

export function getRelatedPaths(paths: NamePath[], direct: string): NamePath[] {
  return paths.filter((path) => path[0] === direct).map(_.tail);
}

/**
 * QueryTree builder
 */

export function buildQueryTree(def: Definition, q: QueryDef): QueryTree {
  const query = queryToQueriable(q);
  const hooks = selectToHooks(q.select).map(({ name, refKey }) => {
    const modelHook = getRef.modelHook(def, refKey);
    return {
      name,
      args: modelHook.args.map(({ name, query }) => {
        // apply a batching filter
        const filter = applyFilterIdInContext(query.fromPath, query.filter);
        // select the __join_connection
        const conn = mkJoinConnection(getRef.model(def, _.first(query.fromPath)!));
        const select = [...query.select, conn];

        return {
          name,
          query: buildQueryTree(def, { ...query, filter, select }),
        };
      }),
      hook: modelHook.hook,
    };
  });
  const model = getRef.model(def, query.retType);

  return {
    name: query.name,
    alias: query.name,
    query,
    queryIdAlias: findTargetIdAlias(q),
    hooks,
    related: queriesFromSelect(def, model, q.select).map((q) => buildQueryTree(def, q)),
  };
}

function queriesFromSelect(def: Definition, model: ModelDef, select: SelectDef): QueryDef[] {
  const deep: NestedSelect[] = kindFilter(select, "nested-select");
  return deep.map((s) => selectToQuery(def, model, s));
}

function selectToQuery(def: Definition, model: ModelDef, select: NestedSelect): QueryDef {
  const ref = getRef(def, select.refKey);
  const namePath = [model.name, ref.name];
  return queryFromParts(
    def,
    select.alias,
    namePath,
    applyFilterIdInContext([model.name], undefined),
    [
      ...transformSelectPath(select.select, _.initial(select.namePath), [model.name]),
      mkJoinConnection(model),
    ]
  );
}

export function transformSelectPath(select: SelectDef, from: string[], to: string[]): SelectDef {
  return select.map((item: SelectItem): SelectItem => {
    return match(item)
      .with({ kind: "expression" }, (item) => ({
        ...item,
        expr: transformExpressionPaths(item.expr, from, to),
      }))
      .with({ kind: "model-hook" }, (item) => ({
        ...item,
        namePath: transformNamePath(item.namePath, from, to),
      }))
      .with({ kind: "nested-select" }, (item) => {
        return {
          ...item,
          namePath: transformNamePath(item.namePath, from, to),
          select: transformSelectPath(item.select, from, to),
        };
      })
      .exhaustive();
  });
}

export function transformNamePath(path: string[], from: string[], to: string[]): string[] {
  ensureEqual(
    _.isEqual(from, _.take(path, from.length)),
    true,
    `Cannot transform name path: ${path.join(".")} doesn't start with ${from.join(".")}`
  );
  return [...to, ..._.drop(path, from.length)];
}

export function transformNamePaths(paths: string[][], from: string[], to: string[]): string[][] {
  return paths.map((path) => transformNamePath(path, from, to));
}

export function transformExpressionPaths(
  exp: TypedExprDef,
  from: string[],
  to: string[]
): TypedExprDef {
  if (exp === undefined) {
    return undefined;
  }
  switch (exp.kind) {
    case "literal":
    case "variable": {
      return exp;
    }
    case "alias": {
      return { ...exp, namePath: transformNamePath(exp.namePath, from, to) };
    }
    case "function": {
      return {
        ...exp,
        args: exp.args.map((arg) => transformExpressionPaths(arg, from, to)),
      };
    }
    case "aggregate-function": {
      return { ...exp, sourcePath: transformNamePath(exp.sourcePath, from, to) };
    }
    case "array": {
      return {
        ...exp,
        elements: exp.elements.map((arg) => transformExpressionPaths(arg, from, to)),
      };
    }
    default: {
      assertUnreachable(exp);
    }
  }
}

function mkJoinConnection(model: ModelDef): SelectableExpression {
  return {
    kind: "expression",
    alias: "__join_connection",
    expr: { kind: "alias", namePath: [model.name, "id"] },
    type: { kind: "integer", nullable: false },
  };
}

export function selectableId(namePath: NamePath): SelectableExpression {
  return {
    kind: "expression",
    alias: GAUDI_INTERNAL_TARGET_ID_ALIAS,
    expr: { kind: "alias", namePath: [...namePath, "id"] },
    type: { kind: "integer", nullable: false },
  };
}
