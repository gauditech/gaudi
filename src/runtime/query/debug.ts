import { QueryDef, SelectFieldItem } from "@src/types/definition";

export function debugQuery(q: QueryDef): void {
  const from = q.fromPath.join(".");
  const fields = q.select
    .filter((s): s is SelectFieldItem => s.kind === "field")
    .map((s) => {
      return debugField(s);
    });
  console.debug(`Querying ${from}: selecting ${fields.join(", ")}`);
}

function debugField(s: SelectFieldItem): string {
  const np = s.namePath.join(".");
  const rk = s.refKey;
  if (s.name === s.alias) {
    return `${np} (${rk})`;
  } else {
    return `${np} as ${s.alias} (${rk})`;
  }
}
