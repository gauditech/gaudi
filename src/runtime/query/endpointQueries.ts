import _ from "lodash";

import {
  QueryTree,
  applyFilterIdInContext,
  buildQueryTree,
  queryFromParts,
  transformSelectPath,
} from "./build";

import { getRef } from "@src/common/refs";
import {
  Definition,
  EndpointDef,
  QueryOrderByAtomDef,
  TargetDef,
  TypedExprDef,
} from "@src/types/definition";

/**
 * Endpoint query builder
 */

export type EndpointQueries = {
  authQueryTree?: QueryTree;
  parentContextQueryTrees: QueryTree[];
  targetQueryTree: QueryTree;
  responseQueryTree: QueryTree;
};

export function buildEndpointQueries(def: Definition, endpoint: EndpointDef): EndpointQueries {
  let authQueryTree;
  if (def.authenticator) {
    const authModel = getRef.model(def, def.authenticator.authUserModel.refKey);
    const filter: TypedExprDef = {
      kind: "function",
      name: "is",
      args: [
        { kind: "alias", namePath: [authModel.refKey, "id"] },
        { kind: "variable", name: "id" },
      ],
    };
    const query = queryFromParts(def, "@auth", [authModel.name], filter, endpoint.authSelect);
    authQueryTree = buildQueryTree(def, query);
  }

  const parentContextQueryTrees = endpoint.parentContext.map((target, index) => {
    const parentTarget = index === 0 ? null : endpoint.parentContext[index - 1];
    const namePath = parentTarget ? [parentTarget.retType, target.name] : [target.retType];
    // apply identifyWith filter
    const targetFilter = targetToFilter({ ...target, namePath });
    // apply filter from it's parent
    const filter = parentTarget
      ? applyFilterIdInContext([parentTarget.retType], targetFilter)
      : targetFilter;

    const select = transformSelectPath(target.select, target.namePath, namePath);

    const query = queryFromParts(def, target.alias, namePath, filter, select);
    return buildQueryTree(def, query);
  });

  // repeat the same for target
  const parentTarget = _.last(endpoint.parentContext);
  const namePath = parentTarget
    ? [parentTarget.retType, endpoint.target.name]
    : [endpoint.target.retType];

  const targetFilter =
    endpoint.kind === "create" || endpoint.kind === "list" || endpoint.kind === "custom-many"
      ? undefined
      : targetToFilter({ ...endpoint.target, namePath });

  const filter = parentTarget
    ? applyFilterIdInContext([parentTarget.retType], targetFilter)
    : targetFilter;

  const select = transformSelectPath(endpoint.target.select, endpoint.target.namePath, namePath);

  const targetQuery = queryFromParts(def, endpoint.target.alias, namePath, filter, select);
  const targetQueryTree = buildQueryTree(def, targetQuery);

  // response query
  const responseQueryTree = buildResponseQueryTree(def, endpoint);

  return { authQueryTree, parentContextQueryTrees, targetQueryTree, responseQueryTree };
}

/**
 * Response query is responsible for fetching the record(s) in order to return the data
 * to the client initiating the request. Response queries ignore `target.identifyWith`
 * filter, instead they fetch by parent `id` (`list`) or a record `id` (`get`, `update`).
 *
 * This is because `id` is the only non-modifiable identifier for a record; it's possible
 * that (custom) actions modify the record in a way that it no longer passes the filter
 * condition. For example, `target.identifyWith` may be a `slug` field and `update` changes
 * the `slug` field to another value.
 *
 * Therefore, we first fetch all targets using `identifyWith` to find their `id` which we
 * use to refetch the `response` query tree after actions are applied.
 */
function buildResponseQueryTree(def: Definition, endpoint: EndpointDef): QueryTree {
  switch (endpoint.kind) {
    case "update":
    case "create":
    case "delete": // FIXME delete should have no response query! Make it nullable?
    case "get":
    case "custom-one": {
      // fetch directly from the table, we have the ID
      const namePath = [endpoint.target.retType];
      const filter = applyFilterIdInContext(namePath, undefined);
      const response = transformSelectPath(
        endpoint.response ?? [],
        endpoint.target.namePath,
        namePath
      );
      const responseQuery = queryFromParts(def, endpoint.target.alias, namePath, filter, response);
      return buildQueryTree(def, responseQuery);
    }
    case "list":
    case "custom-many": {
      const parentTarget = _.last(endpoint.parentContext);
      const namePath = parentTarget
        ? [parentTarget.retType, endpoint.target.name]
        : [endpoint.target.retType];
      const filter = parentTarget
        ? applyFilterIdInContext([parentTarget.retType], undefined)
        : undefined;

      const response = transformSelectPath(
        endpoint.response ?? [],
        endpoint.target.namePath,
        namePath
      );

      // FIXME: Introduce a way to define ordering from the blueprint
      // currently always sorting by id asc
      const orderBy: QueryOrderByAtomDef[] = [
        { exp: { kind: "alias", namePath: [...namePath, "id"] }, direction: "asc" },
      ];
      const responseQuery = queryFromParts(
        def,
        endpoint.target.alias,
        namePath,
        filter,
        response,
        orderBy
      );
      return buildQueryTree(def, responseQuery);
    }
  }
}

function targetToFilter(target: TargetDef): TypedExprDef {
  return {
    kind: "function",
    name: "is",
    args: [
      { kind: "alias", namePath: [...target.namePath, target.identifyWith.name] },
      {
        kind: "variable",
        type: { type: target.identifyWith.type, nullable: false },
        name: target.identifyWith.paramName,
      },
    ],
  };
}
