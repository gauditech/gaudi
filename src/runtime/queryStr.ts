import { source } from "common-tags";
import _ from "lodash";

import { getRef, getTargetModel } from "@src/common/refs";
import { ensureEqual } from "@src/common/utils";
import { BinaryOperator } from "@src/types/ast";
import {
  Definition,
  FilterDef,
  ModelDef,
  QueryDef,
  QueryDefPath,
  SelectableItem,
} from "@src/types/definition";

export function queryToString(def: Definition, q: QueryDef): string {
  switch (q.from.kind) {
    case "model": {
      const { value: model } = getRef<"model">(def, q.from.refKey);
      return source`
      SELECT ${selectToString(def, q.select)}
      FROM "${model.dbname}" AS ${toAlias([model.name])}
      ${q.joinPaths.map((j) => joinToString(def, j))}
      WHERE ${filterToString(q.filter)}`;
    }
    case "query":
      return source`
      SELECT ${selectToString(def, q.select)}
      FROM (${queryToString(def, q.from.query)}) AS ${toAlias(q.from.query.fromPath)}
      ${q.joinPaths.map((j) => joinToString(def, j))}
      WHERE ${filterToString(q.filter)}`;
  }
}

function selectToString(def: Definition, select: SelectableItem[]) {
  return select
    .map((item) => {
      switch (item.kind) {
        case "field": {
          const { value: field } = getRef<"field">(def, item.refKey);
          return `${toAlias(_.initial(item.namePath))}.${field.dbname} AS ${item.alias}`;
        }
        case "constant": {
          // FIXME security issue, this is not an escaped value!!
          ensureEqual(item.type, "integer");
          return `${item.value} AS "${item.alias}"`;
        }
      }
    })
    .join(", ");
}

function toAlias(np: string[]): string {
  return `"${np.join(".")}"`;
}

function filterToString(filter: FilterDef): string {
  if (filter === undefined) return "TRUE = TRUE";
  switch (filter.kind) {
    case "literal": {
      switch (filter.type) {
        case "boolean":
          return filter.value ? "TRUE" : "FALSE";
        case "null":
          return "NULL";
        case "text":
          return `'${filter.value}'`;
        case "integer":
          return filter.value.toString();
      }
    }
    // Due to bug in eslint/prettier, linter complains that `break` is expected in the case "literal"
    // Since inner switch is exaustive, break is unreachable so prettier deletes it
    // eslint-disable-next-line no-fallthrough
    case "alias": {
      const np = filter.namePath.slice(0, filter.namePath.length - 1);
      const f = filter.namePath.at(filter.namePath.length - 1);
      return `${toAlias(np)}.${f}`;
    }
    case "binary": {
      const fstr = `${filterToString(filter.lhs)} ${opToString(filter.operator)} ${filterToString(
        filter.rhs
      )}`;
      if (filter.operator === "or") {
        return `(${fstr})`;
      }
      return fstr;
    }
    case "variable": {
      return `:${filter.name}`;
    }
  }
}

function opToString(op: BinaryOperator): string {
  switch (op) {
    case "is":
      return "=";
    case "is not":
      return "<>";
    default:
      return op.toUpperCase();
  }
}

function getQuerySource(def: Definition, q: QueryDef): ModelDef {
  switch (q.from.kind) {
    case "model":
      return getRef<"model">(def, q.from.refKey).value;
    case "query": {
      return getQuerySource(def, q.from.query);
    }
  }
}

function joinToString(def: Definition, join: QueryDefPath): string {
  const model = getTargetModel(def.models, join.refKey);
  const joinNames = getJoinNames(def, join.refKey);
  const joinMode = join.joinType === "inner" ? "JOIN" : "LEFT JOIN";
  const joinFilter: FilterDef = {
    kind: "binary",
    operator: "is",
    lhs: { kind: "alias", namePath: [..._.initial(join.namePath), joinNames.from] },
    rhs: { kind: "alias", namePath: [...join.namePath, joinNames.to] },
  };

  let src: string;
  if (join.kind === "query") {
    const { value: query } = getRef<"query">(def, join.refKey);
    const sourceModel = getQuerySource(def, query);
    // extend select
    const conn: SelectableItem = {
      kind: "field",
      refKey: `${sourceModel.name}.id`,
      alias: '"__join_connection"',
      name: "id",
      namePath: [sourceModel.name, "id"],
    };
    const retModel = getRef<"model">(def, query.retType).value;
    const fields = retModel.fields.map(
      (f): SelectableItem => ({
        kind: "field",
        refKey: f.refKey,
        alias: f.name,
        name: f.name,
        namePath: [...query.fromPath, f.name],
      })
    );
    src = source`(
      ${queryToString(def, { ...query, select: [...query.select, ...fields, conn] })})`; // FIXME should remove query.select?
  } else {
    src = `"${model.dbname}"`;
  }

  return source`
  ${joinMode}
  ${src} AS ${toAlias(join.namePath)}
  ON ${filterToString(joinFilter)}
  ${join.joinPaths.map((j) => joinToString(def, j))}`;
}

function getJoinNames(def: Definition, refKey: string): { from: string; to: string } {
  const prop = getRef(def, refKey);
  switch (prop.kind) {
    case "reference": {
      const reference = prop.value;
      const { value: field } = getRef<"field">(def, reference.fieldRefKey);
      const { value: refField } = getRef<"field">(def, reference.toModelFieldRefKey);

      return { from: field.name, to: refField.name };
    }
    case "relation": {
      const relation = prop.value;
      const { value: reference } = getRef<"reference">(def, relation.throughRefKey);
      const { value: field } = getRef<"field">(def, reference.fieldRefKey);
      const { value: refField } = getRef<"field">(def, reference.toModelFieldRefKey);

      return { to: field.name, from: refField.name };
    }
    case "query": {
      return { to: '"__join_connection"', from: "id" };
    }
    default:
      throw new Error(`Kind ${prop.kind} not supported`);
  }
}
