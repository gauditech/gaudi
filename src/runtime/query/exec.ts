import { Knex } from "knex";
import _ from "lodash";

import { executeHook } from "../hooks";

import { QueryTree, selectableId } from "./build";
import { queryToString } from "./stringify";

import { Definition, QueryDef } from "@src/types/definition";

type Result = {
  rowCount: number;
  rows: Row[];
};

interface NestedRow {
  id: number;
  [key: string]: string | number | NestedRow[];
}

interface Row {
  id: number;
  [key: string]: string | number;
}

export type Params = Record<string, string | number>;

export async function executeQuery(
  conn: Knex | Knex.Transaction,
  def: Definition,
  q: QueryDef,
  params: Params,
  contextIds: number[]
): Promise<NestedRow[]> {
  const hasId = q.select.find((s) => s.kind === "field" && s.alias === "id");
  let query = q;
  if (!hasId) {
    query = { ...q, select: [...q.select, selectableId(def, query.fromPath)] };
  }
  const sqlTpl = queryToString(def, query).replace(
    ":@context_ids",
    `(${contextIds.map((_, index) => `:context_id_${index}`).join(", ")})`
  );
  const idMap = Object.fromEntries(contextIds.map((id, index) => [`context_id_${index}`, id]));
  const result: Result = await conn.raw(sqlTpl, { ...params, ...idMap });
  return result.rows;
}

export async function executeQueryTree(
  conn: Knex | Knex.Transaction,
  def: Definition,
  qt: QueryTree,
  params: Params,
  contextIds: number[]
): Promise<NestedRow[]> {
  const results = await executeQuery(conn, def, qt.query, params, contextIds);
  if (results.length === 0) return [];
  const resultIds = _.uniq(results.map((r) => r.id));

  // tree
  for (const rel of qt.related) {
    const relResults = await executeQueryTree(conn, def, rel, params, resultIds);
    const groupedById = _.groupBy(relResults, "__join_connection");
    results.forEach((r) => {
      const relResultsForId = (groupedById[r.id] ?? []).map((relR) =>
        _.omit(relR, "__join_connection")
      );
      Object.assign(r, { [rel.name]: relResultsForId });
    });
  }

  results.forEach((result) => {
    qt.hooks.forEach((h) => {
      result[h.name] = executeHook(h.code, {});
    });
  });

  return results;
}
