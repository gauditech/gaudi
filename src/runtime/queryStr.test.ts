import _ from "lodash";

import { compile, compose, parse } from "../index";

import { mkContextQuery, mkTargetQuery } from "./query";
import { queryToString } from "./queryStr";

import { SelectConstantItem } from "@src/types/definition";

describe("queryables", () => {
  it("context query", () => {
    const bp = `
    model Org {
      field slug { type text, unique }
      relation repos { from Repo, through org }
      query public_repos { from repos, filter { is_public is true }}
    }
    model Repo {
      reference org { to Org }
      field is_public { type boolean }
      field name { type text }
    }
    entrypoint Orgs {
      target model Org
      identify with slug
      entrypoint Repos {
        target relation public_repos
        response { id, name }

        get endpoint {}
      }
    }

    `;
    const def = compose(compile(parse(bp)));
    const constant: SelectConstantItem = {
      kind: "constant",
      type: "integer",
      value: 1,
      alias: "exists",
    };
    const endpoint = def.entrypoints[0].entrypoints[0].endpoints[0];
    const q = mkContextQuery(def, endpoint.targets, [
      constant,
      { kind: "field", alias: "id", name: "id", namePath: ["Org", "id"], refKey: "Org.id" },
      {
        kind: "field",
        alias: "name",
        name: "name",
        namePath: ["Org", "public_repos", "name"],
        refKey: "Repo.name",
      },
    ]);
    const s = q ? queryToString(def, q) : "not queryable";
    console.log(s);
  });
  it("target query", () => {
    const bp = `
    model Org {
      field slug { type text, unique }
      relation repos { from Repo, through org }
      query public_repos { from repos, filter { is_public is true }}
    }
    model Repo {
      reference org { to Org }
      field is_public { type boolean }
      field name { type text }
      relation issues { from Issue, through repo }
    }
    model Issue {
      reference repo { to Repo }
      field name { type text }
    }
    entrypoint Orgs {
      target model Org
      identify with slug
      entrypoint Repos {
        target relation public_repos
        response { id, name }

        entrypoint Issues {
          target relation issues
          list endpoint {}
        }
      }
    }

    `;
    const def = compose(compile(parse(bp)));
    const endpoint = def.entrypoints[0].entrypoints[0].entrypoints[0].endpoints[0];
    const q = mkTargetQuery(def, endpoint);
    // console.log(q);
    const s = queryToString(def, q);
    console.log(s);
  });
});
