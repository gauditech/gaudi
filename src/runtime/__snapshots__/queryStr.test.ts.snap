// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queryables context query 1`] = `
"SELECT 1 AS "exists", "Org".id AS id, "Org.public_repos".name AS name
      FROM org AS "Org"
      JOIN
(
      SELECT "Org.repos".id AS id, "Org.repos".is_public AS is_public, "Org.repos".name AS name, "Org.repos".org_id AS org_id, "Org".id AS "__join_connection"
            FROM org AS "Org"
            JOIN
      repo AS "Org.repos"
      ON "Org".id = "Org.repos".org_id
            WHERE "Org.repos".is_public = TRUE) AS "Org.public_repos"
ON "Org".id = "Org.public_repos"."__join_connection"
      WHERE "Org".slug = :org_slug AND "Org.public_repos".id = :repo_id"
`;

exports[`queryables target query 1`] = `
{
  "filter": undefined,
  "from": {
    "kind": "query",
    "query": {
      "filter": {
        "kind": "binary",
        "lhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "slug",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "org_slug",
            "type": "text",
          },
        },
        "operator": "and",
        "rhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "public_repos",
              "id",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "repo_id",
            "type": "integer",
          },
        },
      },
      "from": {
        "kind": "model",
        "refKey": "Org",
      },
      "fromPath": [
        "Org",
        "public_repos",
      ],
      "joinPaths": [
        {
          "joinPaths": [],
          "joinType": "inner",
          "kind": "query",
          "name": "public_repos",
          "namePath": [
            "Org",
            "public_repos",
          ],
          "refKey": "Org.public_repos",
          "retCardinality": "many",
          "retType": "Repo",
        },
      ],
      "name": "Org",
      "nullable": false,
      "refKey": "N/A",
      "retCardinality": "many",
      "retType": "Repo",
      "select": [
        {
          "alias": "id",
          "kind": "field",
          "name": "id",
          "namePath": [
            "Repo",
            "id",
          ],
          "refKey": "Repo.id",
        },
      ],
    },
  },
  "fromPath": [
    "Repo",
    "issues",
  ],
  "joinPaths": [
    {
      "joinPaths": [],
      "joinType": "inner",
      "kind": "relation",
      "name": "issues",
      "namePath": [
        "Org",
        "public_repos",
        "issues",
      ],
      "refKey": "Repo.issues",
      "retCardinality": "many",
      "retType": "Issue",
    },
  ],
  "name": "N/A",
  "nullable": false,
  "refKey": "N/A",
  "retCardinality": "many",
  "retType": "Issue",
  "select": [
    {
      "alias": "id",
      "kind": "field",
      "name": "id",
      "namePath": [
        "Org",
        "public_repos",
        "issues",
        "id",
      ],
      "refKey": "Issue.id",
    },
    {
      "alias": "name",
      "kind": "field",
      "name": "name",
      "namePath": [
        "Org",
        "public_repos",
        "issues",
        "name",
      ],
      "refKey": "Issue.name",
    },
    {
      "alias": "repo_id",
      "kind": "field",
      "name": "repo_id",
      "namePath": [
        "Org",
        "public_repos",
        "issues",
        "repo_id",
      ],
      "refKey": "Issue.repo_id",
    },
  ],
}
`;

exports[`queryables target query 2`] = `
"SELECT "Org.public_repos.issues".id AS id, "Org.public_repos.issues".name AS name, "Org.public_repos.issues".repo_id AS repo_id
      FROM (SELECT "Repo".id AS id       FROM org AS "Org"       JOIN (       SELECT "Org.repos".id AS id, "Org.repos".is_public AS is_public, "Org.repos".name AS name, "Org.repos".org_id AS org_id, "Org".id AS "__join_connection"             FROM org AS "Org"             JOIN       repo AS "Org.repos"       ON "Org".id = "Org.repos".org_id             WHERE "Org.repos".is_public = TRUE) AS "Org.public_repos" ON "Org".id = "Org.public_repos"."__join_connection"       WHERE "Org".slug = :org_slug AND "Org.public_repos".id = :repo_id) AS "Org.public_repos"
      JOIN
issue AS "Org.public_repos.issues"
ON "Org.public_repos".id = "Org.public_repos.issues".repo_id
      WHERE TRUE = TRUE"
`;
